File: lib/reputable_chat/reputation/fingerprint.rb
# frozen_string_literal: true

require "digest"
require_relative "../cryptography/canonical"

module ReputableChat
  module Reputation
    # A hash of the parameters a score was computed under.
    #
    # An attestation publishes the author's own calculated scores as a cache.
    # Reputation is subjective and configuration is per-user, so those numbers
    # are only worth anything to a reader whose parameters match -- the author
    # may run a different k, a different curve, or show_unrated on. Without
    # this fingerprint a reader who used the cache would silently adopt a
    # stranger's settings, and a retune would make every published cache
    # quietly wrong instead of detectably stale.
    #
    # Must agree with public/js/fingerprint.js.
    module Fingerprint
      DOMAIN = "reputablechat:params:v1"

      # Every key that can change a computed score, and nothing else. Seed and
      # session keys are absent deliberately: they cannot move a number, so
      # including them would invalidate caches for no reason.
      SCORING_KEYS = %w[
        precision.scale
        constants.k
        ladder.max_hops
        ladder.max_accounts
        gate.min_rating
        actions.friend.value
        actions.report.value
        vote_curve.cap
        vote_curve.a
        vote_curve.b
        display.visible_above
        display.trusted_at
        display.show_unrated
      ].freeze

      module_function

      def of(config) = digest(values(config))

      def digest(values)
        Digest::SHA256.hexdigest("#{DOMAIN}\n#{Cryptography::Canonical.dump(values)}".b)
      end

      # Values are stringified so that a YAML "0.1" and a JSON 0.1 cannot
      # fingerprint differently for the same setting.
      def values(config)
        SCORING_KEYS.to_h { |key| [key, config.fetch(key).to_s] }
      end
    end
  end
end

File: public/js/fingerprint.js
// A hash of the parameters a score was computed under. Mirrors
// lib/reputable_chat/reputation/fingerprint.rb.
//
// An attestation publishes its author's calculated scores as a cache, and that
// cache only means anything to a reader whose parameters match -- reputation
// is subjective and configuration is per-user. Comparing fingerprints is how a
// reader decides whether to use the cache or do the walk themselves.

import * as canonical from "./canonical.js";

export const DOMAIN = "reputablechat:params:v1";

// Every key that can change a computed score, and nothing else. Seed and
// session keys are absent deliberately: they cannot move a number, so
// including them would invalidate caches for no reason.
export const SCORING_KEYS = [
  "precision.scale",
  "constants.k",
  "ladder.max_hops",
  "ladder.max_accounts",
  "gate.min_rating",
  "actions.friend.value",
  "actions.report.value",
  "vote_curve.cap",
  "vote_curve.a",
  "vote_curve.b",
  "display.visible_above",
  "display.trusted_at",
  "display.show_unrated",
];

// Values are stringified so a YAML "0.1" and a JSON 0.1 cannot fingerprint
// differently for the same setting.
export function values(lookup) {
  const out = {};
  for (const key of SCORING_KEYS) out[key] = String(lookup(key));
  return out;
}

export async function of(lookup) {
  return digest(values(lookup));
}

export async function digest(map) {
  const input = new TextEncoder().encode(`${DOMAIN}\n${canonical.dump(map)}`);
  const hashed = await crypto.subtle.digest("SHA-256", input);
  return [...new Uint8Array(hashed)].map((b) => b.toString(16).padStart(2, "0")).join("");
}
